lenskit-core/src/main/java/org/lenskit/data/entities/EntityBuilder.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2014 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.data.entities;
+
+/**
+ * Base class for entity builders.  The standard implementation is {@link BasicEntityBuilder}, which can be created
+ * with {@link Entities#newBuilder()}.
+ */
+public abstract class EntityBuilder {
+    protected EntityType type;
+    protected long id;
+    protected boolean idSet;
+
+    /**
+     * Construct a new entity builder.
+     *
+     * **Note:** This constructor calls {@link #setId(long)} and {@link #setType(EntityType)}, so if those methods
+     * are overloaded, they should be able to run without the derived class constructor having finished.
+     *
+     * @param initId The initial ID.
+     * @param initIdSet Whether the ID is initially set.
+     * @param typ The entity type.
+     */
+    protected EntityBuilder(long initId, boolean initIdSet, EntityType typ) {
```

Comment:
It is probably needless complexity.

I don't know that we have a use case for `type` being mutable. I can make it immutable and we can change that later if we have a reason.